### PR TITLE
Make unknownFields nullable to reflection

### PR DESCRIPTION
**Pull request by swankjesse**
This is particularly unfortunate but the least worst of several
bad options. Our backends incorrectly use Gson to reflectively
create Wire objects, those Wire objects don't include a good
representation of unknownFields.

### CODE REVIEW

File: wire-grpc-tests/src/test/proto-grpc/routeguide/Feature.kt [hunk 1]
```diff
@@ -55,7 +55,7 @@ class Feature(
   override fun equals(other: Any?): Boolean {
     if (other === this) return true
     if (other !is Feature) return false
-    return unknownFields == other.unknownFields
+    return unknownFields() == other.unknownFields()
```
Comment (swankjesse) [inline wire-grpc-tests/src/test/proto-grpc/routeguide/Feature.kt]: no longer a val. I couldn’t figure out how to have a nullable JVM field named `unknownFields` and also a non-nullable val with the same name.

File: wire-gson-support/src/test/java/com/squareup/dinosaurs/Dinosaur.java [hunk 2]
```diff
@@ -0,0 +1,252 @@
+// Code generated by Wire protocol buffer compiler, do not edit.
```
Comment (swankjesse) [inline wire-gson-support/src/test/java/com/squareup/dinosaurs/Dinosaur.java]: Just for the regression test.

File: wire-runtime/src/commonMain/kotlin/com/squareup/wire/Message.kt [hunk 3]
```diff
@@ -30,7 +30,7 @@ expect abstract class Message<M : Message<M, B>, B : Message.Builder<M, B>> prot
    * Returns a byte string containing the proto encoding of this message's unknown fields. Returns
    * an empty byte string if this message has no unknown fields.
    */
-  val unknownFields: ByteString
+  fun unknownFields(): ByteString
```
Comment (JakeWharton) [inline wire-runtime/src/commonMain/kotlin/com/squareup/wire/Message.kt]: This shouldn't have to change to support the behavior.

File: wire-grpc-tests/src/test/proto-grpc/routeguide/Feature.kt (hunk 1 above)
Comment (JakeWharton) [reply wire-grpc-tests/src/test/proto-grpc/routeguide/Feature.kt]: You probably need to do something like
```kotlin
@JvmName("unknownFields")
private val _unknownFields: ByteString?

val unknownFields get() = _unknownFields ?: EMPTY
```

Comment (swankjesse) [reply wire-grpc-tests/src/test/proto-grpc/routeguide/Feature.kt]: I tried it and IntelliJ complained. Lemme try again.

Comment (swankjesse) [reply wire-grpc-tests/src/test/proto-grpc/routeguide/Feature.kt]: Yeah, `@JvmName` can’t target field names.

```
/**
 * Specifies the name for the Java class or method which is generated from this element.
 *
 * See the [Kotlin language documentation](https://kotlinlang.org/docs/reference/java-to-kotlin-interop.html#handling-signature-clashes-with-jvmname)
 * for more information.
 * @property name the name of the element.
 */
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER, AnnotationTarget.FILE)
@Retention(AnnotationRetention.BINARY)
@MustBeDocumented
public actual annotation class JvmName(actual val name: String)
```

File: wire-runtime/src/commonMain/kotlin/com/squareup/wire/Message.kt (hunk 3 above)
Comment (swankjesse) [reply wire-runtime/src/commonMain/kotlin/com/squareup/wire/Message.kt]: Lets try something completely different:
https://github.com/square/wire/pull/1160/files

**Pull request closed without merging**